modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * Ignite SQL call rewrite table. Performs unconditional rewrites for some predefined Calcite SQL operators,
+ * which can't be extended other ways by Ignite.
+ */
+public class IgniteSqlCallRewriteTable {
+    /** Instance. */
+    public static final IgniteSqlCallRewriteTable INSTANCE = new IgniteSqlCallRewriteTable();
+
+    /** Registered rewriters map. */
+    private final Map<SqlOperator, BiFunction<SqlValidator, SqlCall, SqlCall>> map = new HashMap<>();
+
+    /** */
+    private IgniteSqlCallRewriteTable() {
+        register(SqlLibraryOperators.NVL, IgniteSqlCallRewriteTable::nvlRewriter);
+        register(SqlLibraryOperators.DECODE, IgniteSqlCallRewriteTable::decodeRewriter);
+    }
+
+    /** Registers rewriter for SQL operator. */
+    public void register(SqlOperator operator, BiFunction<SqlValidator, SqlCall, SqlCall> rewriter) {
+        map.put(operator, rewriter);
+    }
+
+    /** Rewrites SQL call. */
+    SqlCall rewrite(SqlValidator validator, SqlCall call) {
+        BiFunction<SqlValidator, SqlCall, SqlCall> rewriter = map.get(call.getOperator());
+
+        return rewriter == null ? call : rewriter.apply(validator, call);
+    }
+
+    /** Rewrites NVL call to CASE WHEN call. */
+    public static SqlCall nvlRewriter(SqlValidator validator, SqlCall call) {
+        validateQuantifier(validator, call); // check DISTINCT/ALL
+
+        List<SqlNode> operands = call.getOperandList();
+
+        if (operands.size() == 2) {
+            SqlParserPos pos = call.getParserPosition();
+
+            SqlNodeList whenList = new SqlNodeList(pos);
+            SqlNodeList thenList = new SqlNodeList(pos);
+
+            whenList.add(SqlStdOperatorTable.IS_NOT_NULL.createCall(pos, operands.get(0)));
+            thenList.add(SqlNode.clone(operands.get(0)));
+            SqlNode elseExpr = operands.get(1);
+
+            return SqlCase.createSwitched(pos, null, whenList, thenList, elseExpr);
+        }
+        else
+            return call; // Operands count will be validated and exception will be thrown later.
+    }
+
+    /** Rewrites DECODE call to CASE WHEN call. */
+    public static SqlCall decodeRewriter(SqlValidator validator, SqlCall call) {
+        validateQuantifier(validator, call); // check DISTINCT/ALL
+
+        List<SqlNode> operands = call.getOperandList();
+
+        SqlParserPos pos = call.getParserPosition();
+
+        SqlNode op0 = operands.get(0);
+
+        SqlNodeList whenList = new SqlNodeList(pos);
+        SqlNodeList thenList = new SqlNodeList(pos);
+
+        for (int i = 1; i < operands.size() - 1; i += 2) {
+            whenList.add(SqlStdOperatorTable.IS_NOT_DISTINCT_FROM.createCall(pos, SqlNode.clone(op0), operands.get(i)));
```

Comment:
Do we need `SqlNode.clone(op0)` here and above?

Reply:
Yes, we need to guarantee that oriented graph is a tree, or we can get some unexpected behaviour during recursive tree traversal.